lenskit-api/src/main/java/org/lenskit/api/ItemRecommender.java

```diff
@@ -0,0 +1,133 @@
+/*
+ * LensKit, an open source recommender systems toolkit.
+ * Copyright 2010-2014 LensKit Contributors.  See CONTRIBUTORS.md.
+ * Work on LensKit has been funded by the National Science Foundation under
+ * grants IIS 05-34939, 08-08692, 08-12148, and 10-17697.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program; if not, write to the Free Software Foundation, Inc., 51
+ * Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+package org.lenskit.api;
+
+import javax.annotation.Nullable;
+import java.util.Set;
+
+/**
+ * Interface for recommending items. Several methods are provided, of varying
+ * generality.
+ *
+ * The core idea of the recommend API is to recommend <i>n</i> items for a user,
+ * where the items recommended are taken from a set of candidate items and
+ * further constrained by an exclude set of forbidden items. Items in the
+ * candidate set but not in the exclude set are considered viable for
+ * recommendation.
+ *
+ * ## Candidate Items
+ *
+ * By default, the candidate set is the universe of all items the recommender
+ * knows about. The default exclude set is somewhat more subtle. Its exact
+ * definition varies across implementations, but will be the set of items the
+ * system believes the user will not be interested in by virtue of already
+ * having or knowing about them. For example, rating-based recommenders will
+ * exclude the items the user has rated, and purchase-based recommenders will
+ * typically exclude items the user has purchased. Some implementations may
+ * allow this to be configured. Client code always has the option of manually
+ * specifying the exclude set, however, so applications with particular needs in
+ * this respect can manually provide the sets they need respected.
+ *
+ * ## Ordering
+ *
+ * If the recommender has an opinion about the order in which recommendations should be displayed,
+ * the result set will present items in that order.  For many recommenders, this will be descending order
+ * by score; however, this interface does not guarantee a relationship between scores and ordering.
+ *
+ * @compat Public
+ */
+public interface ItemRecommender {
```

Comment:
I assume we will also have an abstract class that does some aggressive delegation to make most people only need to implement one function?